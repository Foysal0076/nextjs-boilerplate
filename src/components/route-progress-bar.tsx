'use client'

import NextTopLoader from 'nextjs-toploader'

export default function RouteProgressBar() {
  return (
    <NextTopLoader
      color='hsl(var(--primary))'
      showSpinner={false}
      crawlSpeed={100}
      speed={100}
    />
  )
}
