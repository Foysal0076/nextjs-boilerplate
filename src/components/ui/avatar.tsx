import Image from 'next/image'

import { cn, createInitial } from '@/utils'

type Props = {
  name: string
  url?: string
  size?: number
  className?: string
  textClassName?: string
}

const Avatar = ({ name, url, className, textClassName }: Props) => {
  const initial = createInitial(name)

  return (
    <div
      data-testid='avatar'
      className={cn(
        `relative flex h-8 w-8 items-center justify-center rounded-full border bg-card text-sm font-bold text-card-foreground`,
        className
      )}>
      {url ? (
        <Image
          src={url}
          alt={name}
          fill
          className='rounded-full object-cover'
          sizes='(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw'
          placeholder='blur'
          blurDataURL={`_next/image?url=${url}&w=16&q=1`}
        />
      ) : (
        <span className={cn('text-inherit', textClassName)}>{initial}</span>
      )}
    </div>
  )
}
Avatar.displayName = 'Avatar'

export { Avatar }
