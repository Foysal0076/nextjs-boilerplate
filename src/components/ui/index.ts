export { Button } from './button'
export { Input } from './input'
export { Textarea } from './textarea'
export { Avatar } from './avatar'
export { Spinner } from './spinner'
