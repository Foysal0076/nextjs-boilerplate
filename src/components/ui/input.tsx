import { forwardRef, InputHTMLAttributes, ReactNode } from 'react'

import { cn } from '@/utils'

export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  label?: string
  labelClassName?: string
  error?: string
  helperText?: string
  startAdornment?: {
    adornment: ReactNode
    onClick?: () => void
    className?: string
  }
  endAdornment?: {
    adornment: ReactNode
    onClick?: () => void
    className?: string
  }
}

const Input = forwardRef<HTMLInputElement, InputProps>(
  (
    {
      className,
      type,
      label,
      disabled,
      error,
      labelClassName,
      helperText,
      startAdornment,
      endAdornment,
      ...props
    },
    ref
  ) => {
    return (
      <div className='relative flex flex-col gap-1.5'>
        {label && (
          <label
            htmlFor={props.id}
            className={cn(
              'text-sm font-medium',
              {
                'text-destructive': error,
                'after:ml-0.5 after:text-destructive after:content-["*"]':
                  props.required,
                'text-muted-foreground opacity-50': disabled,
              },
              labelClassName
            )}>
            {label}
          </label>
        )}
        <div className='relative'>
          <input
            type={type}
            className={cn(
              'flex h-9 w-full rounded-md border border-input bg-transparent px-3 py-1 text-sm shadow-sm transition-colors',
              'file:border-0 file:bg-transparent file:text-sm file:font-medium',
              'placeholder:text-muted-foreground',
              'focus-visible:outline-none',
              'disabled:cursor-not-allowed disabled:opacity-50',
              {
                'pl-9': startAdornment,
                'pr-9': endAdornment,
                'focus-visible:ring-1 focus-visible:ring-ring': !error,
                'border-destructive ring-destructive': error,
              },
              className
            )}
            ref={ref}
            disabled={disabled}
            {...props}
          />
          {startAdornment && (
            <div
              onClick={startAdornment.onClick}
              className={cn(
                'absolute left-2.5 top-1/2 -translate-y-1/2 text-muted-foreground',
                { 'cursor-pointer': startAdornment.onClick },
                startAdornment.className
              )}>
              {startAdornment.adornment}
            </div>
          )}
          {endAdornment && (
            <div
              onClick={endAdornment.onClick}
              className={cn(
                'absolute right-2.5 top-1/2 -translate-y-1/2 text-muted-foreground',
                { 'cursor-pointer': endAdornment.onClick },
                endAdornment.className
              )}>
              {endAdornment.adornment}
            </div>
          )}
        </div>
        {(helperText || error) && (
          <span
            className={cn(
              'text-xs',
              error ? 'text-destructive' : 'text-muted-foreground'
            )}>
            {error || helperText}
          </span>
        )}
      </div>
    )
  }
)
Input.displayName = 'Input'

export { Input }
