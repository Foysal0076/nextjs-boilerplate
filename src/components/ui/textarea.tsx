import { forwardRef, TextareaHTMLAttributes } from 'react'

import { cn } from '@/utils'

export interface TextareaProps
  extends TextareaHTMLAttributes<HTMLTextAreaElement> {
  label?: string
  labelClassName?: string
  error?: string
  helperText?: string
}

const Textarea = forwardRef<HTMLTextAreaElement, TextareaProps>(
  (
    { className, label, disabled, error, labelClassName, helperText, ...props },
    ref
  ) => {
    return (
      <div className='relative flex flex-col gap-1.5'>
        {label && (
          <label
            htmlFor={props.id}
            className={cn(
              'text-sm font-medium',
              {
                'text-destructive': error,
                'after:ml-0.5 after:text-destructive after:content-["*"]':
                  props.required,
                'text-muted-foreground opacity-50': disabled,
              },
              labelClassName
            )}>
            {label}
          </label>
        )}
        <textarea
          className={cn(
            'flex min-h-[60px] w-full rounded-md border border-input bg-transparent px-3 py-2 text-sm shadow-sm transition-colors',
            'placeholder:text-muted-foreground',
            'focus-visible:outline-none',
            'disabled:cursor-not-allowed disabled:opacity-50',
            {
              'focus-visible:ring-1 focus-visible:ring-ring': !error,
              'border-destructive ring-destructive': error,
            },
            className
          )}
          ref={ref}
          disabled={disabled}
          {...props}
        />
        {(helperText || error) && (
          <span
            className={cn(
              'text-xs',
              error ? 'text-destructive' : 'text-muted-foreground'
            )}>
            {error || helperText}
          </span>
        )}
      </div>
    )
  }
)
Textarea.displayName = 'Textarea'

export { Textarea }
