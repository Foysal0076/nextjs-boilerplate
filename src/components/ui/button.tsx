import { cva, type VariantProps } from 'class-variance-authority'
import { ButtonHTMLAttributes, forwardRef } from 'react'

import { Spinner } from '@/components/ui'
import { cn } from '@/utils'

const buttonVariants = cva(
  'inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 active:scale-95 transition-transform',
  {
    variants: {
      variant: {
        default:
          'bg-primary text-primary-foreground shadow hover:bg-primary/90',
        destructive:
          'bg-destructive text-destructive-foreground shadow-sm hover:bg-destructive/90',
        outline:
          'border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground',
        secondary:
          'bg-secondary text-secondary-foreground shadow-sm hover:bg-secondary/80',
        ghost: 'hover:bg-accent hover:text-accent-foreground',
        link: 'text-primary underline-offset-4 hover:underline',
      },
      size: {
        default: 'h-9 px-4 py-2',
        sm: 'h-8 rounded-md px-3 text-xs',
        lg: 'h-10 rounded-md px-8',
        icon: 'h-9 w-9',
      },
    },
    defaultVariants: {
      variant: 'default',
      size: 'default',
    },
  }
)

export interface ButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants> {
  loading?: boolean
}

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    { className, variant, size, loading, children, disabled, ...props },
    ref
  ) => {
    return (
      <button
        className={cn('relative', buttonVariants({ variant, size, className }))}
        ref={ref}
        disabled={disabled || loading}
        {...props}>
        {loading && (
          <div
            className='absolute inset-0 flex items-center justify-center'
            data-testid='spinner'>
            <Spinner className='text-foreground' />
          </div>
        )}
        <div
          className={cn(`visible`, {
            invisible: loading,
            'opacity-0': loading,
          })}>
          {children}
        </div>
      </button>
    )
  }
)
Button.displayName = 'Button'

export { Button }
