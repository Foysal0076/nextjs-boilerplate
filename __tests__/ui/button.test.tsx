import '@testing-library/jest-dom'

import { fireEvent, render, screen } from '@testing-library/react'

import { Button } from '@/components/ui'

describe('Button', () => {
  it('renders correctly with default props', () => {
    render(<Button>Click me</Button>)
    const button = screen.getByRole('button', { name: /click me/i })
    expect(button).toBeInTheDocument()
    expect(button).toHaveClass(
      'bg-primary text-primary-foreground shadow hover:bg-primary/90'
    )
  })

  it('applies custom className', () => {
    render(<Button className='custom-class'>Custom Button</Button>)
    const button = screen.getByRole('button', {
      name: 'Custom Button',
    })
    expect(button)
  })

  it('handles click events', async () => {
    const handleClick = jest.fn()

    render(<Button onClick={handleClick}>Click me</Button>)

    fireEvent.click(screen.getByRole('button'))
    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  it('can be disabled', () => {
    render(<Button disabled>Disabled Button</Button>)
    const button = screen.getByRole('button', { name: 'Disabled Button' })
    expect(button).toBeDisabled()
  })

  it('forwards ref correctly', () => {
    const ref = jest.fn()
    render(<Button ref={ref}>Button</Button>)
    expect(ref).toHaveBeenCalled()
  })

  //snapshot tests
  it('renders all variants and sizes', () => {
    const { container } = render(
      <>
        {/* Variants */}
        <Button variant='default'>Default</Button>
        <Button variant='destructive'>Destructive</Button>
        <Button variant='outline'>Outline</Button>
        <Button variant='secondary'>Secondary</Button>
        <Button variant='ghost'>Ghost</Button>
        <Button variant='link'>Link</Button>

        {/* Sizes */}
        <Button size='default'>Default Size</Button>
        <Button size='sm'>Small</Button>
        <Button size='lg'>Large</Button>
        <Button size='icon'>Icon</Button>
      </>
    )
    expect(container).toMatchSnapshot()
  })
})
