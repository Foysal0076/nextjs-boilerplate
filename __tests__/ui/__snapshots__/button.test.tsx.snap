// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders all variants and sizes 1`] = `
<div>
  <button
    class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 active:scale-95 transition-transform bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
  >
    <div
      class="visible"
    >
      Default
    </div>
  </button>
  <button
    class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 active:scale-95 transition-transform bg-destructive text-destructive-foreground shadow-sm hover:bg-destructive/90 h-9 px-4 py-2"
  >
    <div
      class="visible"
    >
      Destructive
    </div>
  </button>
  <button
    class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 active:scale-95 transition-transform border border-input bg-background shadow-sm hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2"
  >
    <div
      class="visible"
    >
      Outline
    </div>
  </button>
  <button
    class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 active:scale-95 transition-transform bg-secondary text-secondary-foreground shadow-sm hover:bg-secondary/80 h-9 px-4 py-2"
  >
    <div
      class="visible"
    >
      Secondary
    </div>
  </button>
  <button
    class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 active:scale-95 transition-transform hover:bg-accent hover:text-accent-foreground h-9 px-4 py-2"
  >
    <div
      class="visible"
    >
      Ghost
    </div>
  </button>
  <button
    class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 active:scale-95 transition-transform text-primary underline-offset-4 hover:underline h-9 px-4 py-2"
  >
    <div
      class="visible"
    >
      Link
    </div>
  </button>
  <button
    class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 active:scale-95 transition-transform bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 px-4 py-2"
  >
    <div
      class="visible"
    >
      Default Size
    </div>
  </button>
  <button
    class="relative inline-flex items-center justify-center whitespace-nowrap font-medium focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 active:scale-95 transition-transform bg-primary text-primary-foreground shadow hover:bg-primary/90 h-8 rounded-md px-3 text-xs"
  >
    <div
      class="visible"
    >
      Small
    </div>
  </button>
  <button
    class="relative inline-flex items-center justify-center whitespace-nowrap text-sm font-medium focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 active:scale-95 transition-transform bg-primary text-primary-foreground shadow hover:bg-primary/90 h-10 rounded-md px-8"
  >
    <div
      class="visible"
    >
      Large
    </div>
  </button>
  <button
    class="relative inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:pointer-events-none disabled:opacity-50 active:scale-95 transition-transform bg-primary text-primary-foreground shadow hover:bg-primary/90 h-9 w-9"
  >
    <div
      class="visible"
    >
      Icon
    </div>
  </button>
</div>
`;
